src/handler/mod.rs

```diff
@@ -998,17 +1111,44 @@ impl Handler {
     /// Nodes response.
     async fn handle_response(&mut self, node_address: NodeAddress, response: Response) {
         // Find a matching request, if any
-        if let Some(mut request_call) = self.active_requests.remove(&node_address) {
+        trace!("Received {} response", response.body);
+
+        let (request_call, is_regconf) =
+            if let Some(request_call) = self.active_requests_regconf.remove(&node_address) {
+                (Some(request_call), true)
+            } else {
+                (self.active_requests.remove(&node_address), false)
+            };
+
+        if let Some(mut request_call) = request_call {
             if request_call.id() != &response.id {
-                trace!(
-                    "Received an RPC Response to an unknown request. Likely late response. {}",
-                    node_address
-                );
                 // add the request back and reset the timer
-                self.active_requests.insert(node_address, request_call);
+                if is_regconf {
+                    trace!(
+                        "Received an RPC Response from a node we are also waiting for a REGISTERCONFIRMATION from. {}",
+                        node_address
+                    );
+                    self.active_requests_regconf
+                        .insert(node_address, request_call);
+                } else {
+                    trace!(
+                        "Received an RPC Response to an unknown request. Likely late response. {}",
+                        node_address
+                    );
+                    self.active_requests.insert(node_address, request_call);
+                }
                 return;
             }
 
+            let blacklist_peer = |handler: &mut Handler| {
+                // Remove the expected response
+                handler.remove_expected_response(node_address.socket_addr);
```

Comment:
Do we want to fail the session also?

Reply:
If you would please show me how to make an alternative to an async closure in stable Rust yes sure we want to add ```fail_request``` (which then also fails session) into that scope which is otherwise always called before ```blacklist_peer``` is called